### PR TITLE
Added dispatcher process

### PR DESCRIPTION
**Pull request by jshaofuturewei**

### CODE REVIEW

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	dispatcherSelector := fields.ParseSelectorOrDie(
+		",metatdata.namespace" + p.namespace + ",metatdata.name=" + p.name)
+	dispatcherLW := cache.NewListWatchFromClient(p.dispatcherClientset, "Dispatchers", metav1.NamespaceAll, dispatcherSelector)
+
+	dispatcherInformer := cache.NewSharedIndexInformer(dispatcherLW, &dispatcherv1.Dispatcher{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+
+	dispatcherInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			quit <- struct{}{}
+		},
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:123]: Please test to make sure that the dispatcher process is gracefully deleted through this "quit channel signalling" mechanism

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+			if !ok {
+				klog.Warningf("Failed to convert a old object  %+v to a dispatcher", old)
+				return
+			}
+			newDispatcher, ok := new.(*dispatcherv1.Dispatcher)
+			if !ok {
+				klog.Warningf("Failed to convert a new object  %+v to a dispatcher", new)
+				return
+			}
+			if reflect.DeepEqual(oldDispatcher.Spec.POD, newDispatcher.Spec.POD) {
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:135]: change "equal" to "not equal" to process the dispatcher update

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+				podInformer := p.initClusterInformers(p.resetCh, newDispatcher.Spec.POD)
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:136]: need to modify this to do a "filtered watch" on the POD whose status==binded and assigend cluster falls into this dispatcher' cluster list

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+	clusterInformer := p.initClusterInformers(p.resetCh, p.clusterIds)
+	podInformer := p.initPodInformers(p.resetCh, p.clusterIds)
+	go clusterInformer.Run(p.resetCh)
+	go podInformer.Run(p.resetCh)
+	go dispatcherInformer.Run(quit)
+	wait.Until(p.ScheduleOne, 0, quit)
+}
+
+func (p *Process) initClusterInformers(resetCh chan struct{}, clusterIds []string) cache.SharedIndexInformer {
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:150]: These cluster related informer codes are not needed. You can remove them

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+func (p *Process) initPodInformers(resetCh chan struct{}, clusterIds []string) cache.SharedIndexInformer {
+	resetCh <- struct{}{}
+	conditions := "metedata.namespace=" + p.namespace + ","
+	for _, clusterId := range clusterIds {
+		conditions = conditions + "spec.clusterName=" + clusterId + ";"
+	}
+	clusterSelector := fields.ParseSelectorOrDie(conditions)
+
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:172]: You need to add another selector/filter "POD status=binded" or "POD status=deleting" to the POD watch

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+	lw := cache.NewListWatchFromClient(p.clientset.CoreV1(), "Pods", metav1.NamespaceAll, clusterSelector)
+
+	podInformer := cache.NewSharedIndexInformer(lw, &v1.Pod{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		//TODO
+		DeleteFunc: func(obj interface{}) {
+			//Here we can get all the pods information and compose openstack request
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:179]: add codes here to invoke openstack "servercreate" and "deleteInstance"

File: globalscheduler/controllers/dispatcher/dispatcher_process.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+	for _, item := range list {
+		if item.String() == value {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Process) ScheduleOne() {
+
```
Comment (cathyhongzhang) [inline globalscheduler/controllers/dispatcher/dispatcher_process.go:210]: suggest to change the name to "SendPodToCluster"

**Review by cathyhongzhang — APPROVED** [review]

File: globalscheduler/controllers/dispatcher/dispatcher_process.go (hunk 7 above)
Comment (kimeunju108) [reply globalscheduler/controllers/dispatcher/dispatcher_process.go:197]: This hasn't applied yet

**Review by kimeunju108 — APPROVED** [review]
I left a comment because there is one Cathy's review hasn't applied.

File: globalscheduler/controllers/dispatcher/dispatcher_process.go (hunk 7 above)
Comment (jshaofuturewei) [reply globalscheduler/controllers/dispatcher/dispatcher_process.go:197]: Applied

**Pull request merged**